<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.4.555">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Alexandra Mpekou">
<meta name="dcterms.date" content="2024-08-08">

<title>Analyzing Biological Networks</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
  margin-bottom: 0em;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>


<script src="BioNet_files/libs/clipboard/clipboard.min.js"></script>
<script src="BioNet_files/libs/quarto-html/quarto.js"></script>
<script src="BioNet_files/libs/quarto-html/popper.min.js"></script>
<script src="BioNet_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="BioNet_files/libs/quarto-html/anchor.min.js"></script>
<link href="BioNet_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="BioNet_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="BioNet_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="BioNet_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="BioNet_files/libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<style>

      .quarto-title-block .quarto-title-banner h1,
      .quarto-title-block .quarto-title-banner h2,
      .quarto-title-block .quarto-title-banner h3,
      .quarto-title-block .quarto-title-banner h4,
      .quarto-title-block .quarto-title-banner h5,
      .quarto-title-block .quarto-title-banner h6
      {
        color: #6c9685;
      }

      .quarto-title-block .quarto-title-banner {
        color: #6c9685;
      }
</style>

  <script>window.backupDefine = window.define; window.define = undefined;</script><script src="https://cdn.jsdelivr.net/npm/katex@0.15.1/dist/katex.min.js"></script>
  <script>document.addEventListener("DOMContentLoaded", function () {
 var mathElements = document.getElementsByClassName("math");
 var macros = [];
 for (var i = 0; i < mathElements.length; i++) {
  var texText = mathElements[i].firstChild;
  if (mathElements[i].tagName == "SPAN") {
   katex.render(texText.data, mathElements[i], {
    displayMode: mathElements[i].classList.contains('display'),
    throwOnError: false,
    macros: macros,
    fleqn: false
   });
}}});
  </script>
  <script>window.define = window.backupDefine; window.backupDefine = undefined;</script><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/katex@0.15.1/dist/katex.min.css">

<script type="text/javascript">
const typesetMath = (el) => {
  if (window.MathJax) {
    // MathJax Typeset
    window.MathJax.typeset([el]);
  } else if (window.katex) {
    // KaTeX Render
    var mathElements = el.getElementsByClassName("math");
    var macros = [];
    for (var i = 0; i < mathElements.length; i++) {
      var texText = mathElements[i].firstChild;
      if (mathElements[i].tagName == "SPAN") {
        window.katex.render(texText.data, mathElements[i], {
          displayMode: mathElements[i].classList.contains('display'),
          throwOnError: false,
          macros: macros,
          fleqn: false
        });
      }
    }
  }
}
window.Quarto = {
  typesetMath
};
</script>

<link rel="stylesheet" href="styles.css">
</head>

<body>

<header id="title-block-header" class="quarto-title-block default toc-left page-columns page-full">
  <div class="quarto-title-banner page-columns page-full">
    <div class="quarto-title column-body">
      <div class="quarto-title-block"><div><h1 class="title">Analyzing Biological Networks</h1><button type="button" class="btn code-tools-button dropdown-toggle" id="quarto-code-tools-menu" data-bs-toggle="dropdown" aria-expanded="false"><i class="bi"></i> Code</button><ul class="dropdown-menu dropdown-menu-end" aria-labelelledby="quarto-code-tools-menu"><li><a id="quarto-show-all-code" class="dropdown-item" href="javascript:void(0)" role="button">Show All Code</a></li><li><a id="quarto-hide-all-code" class="dropdown-item" href="javascript:void(0)" role="button">Hide All Code</a></li><li><hr class="dropdown-divider"></li><li><a id="quarto-view-source" class="dropdown-item" href="javascript:void(0)" role="button" data-quarto-source-url="https://github.com/darwinsorchid/Biological-Networks">View Source</a></li></ul></div></div>
            <p class="subtitle lead">Protein-Protein Interaction Network</p>
                      </div>
  </div>
    
  
  <div class="quarto-title-meta">

      <div>
      <div class="quarto-title-meta-heading">Author</div>
      <div class="quarto-title-meta-contents">
               <p>Alexandra Mpekou </p>
            </div>
    </div>
      
      <div>
      <div class="quarto-title-meta-heading">Published</div>
      <div class="quarto-title-meta-contents">
        <p class="date">August 8, 2024</p>
      </div>
    </div>
    
      
    </div>
    
  
  </header><div id="quarto-content" class="page-columns page-rows-contents page-layout-article toc-left">
<div id="quarto-sidebar-toc-left" class="sidebar toc-left">
  <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Contents</h2>
   
  <ul>
  <li><a href="#introduction" id="toc-introduction" class="nav-link active" data-scroll-target="#introduction"><span class="header-section-number">1</span> Introduction</a></li>
  <li><a href="#results" id="toc-results" class="nav-link" data-scroll-target="#results"><span class="header-section-number">2</span> Results</a>
  <ul>
  <li><a href="#graph-visualization" id="toc-graph-visualization" class="nav-link" data-scroll-target="#graph-visualization"><span class="header-section-number">2.1</span> Graph Visualization</a></li>
  <li><a href="#node-degree-k" id="toc-node-degree-k" class="nav-link" data-scroll-target="#node-degree-k"><span class="header-section-number">2.2</span> Node Degree (k)</a>
  <ul class="collapse">
  <li><a href="#degree-distribution-pk" id="toc-degree-distribution-pk" class="nav-link" data-scroll-target="#degree-distribution-pk"><span class="header-section-number">2.2.1</span> Degree distribution (P(k))</a></li>
  </ul></li>
  <li><a href="#clustering-and-communities" id="toc-clustering-and-communities" class="nav-link" data-scroll-target="#clustering-and-communities"><span class="header-section-number">2.3</span> Clustering and Communities</a>
  <ul class="collapse">
  <li><a href="#community-detection" id="toc-community-detection" class="nav-link" data-scroll-target="#community-detection"><span class="header-section-number">2.3.1</span> Community Detection</a></li>
  <li><a href="#modularity" id="toc-modularity" class="nav-link" data-scroll-target="#modularity"><span class="header-section-number">2.3.2</span> Modularity</a></li>
  <li><a href="#betweenness-centrality" id="toc-betweenness-centrality" class="nav-link" data-scroll-target="#betweenness-centrality"><span class="header-section-number">2.3.3</span> Betweenness Centrality</a></li>
  <li><a href="#transitivity" id="toc-transitivity" class="nav-link" data-scroll-target="#transitivity"><span class="header-section-number">2.3.4</span> Transitivity</a></li>
  </ul></li>
  <li><a href="#the-importance-of-hubs" id="toc-the-importance-of-hubs" class="nav-link" data-scroll-target="#the-importance-of-hubs"><span class="header-section-number">2.4</span> The Importance Of Hubs</a></li>
  </ul></li>
  <li><a href="#conclusion" id="toc-conclusion" class="nav-link" data-scroll-target="#conclusion"><span class="header-section-number">3</span> Conclusion</a></li>
  </ul>
</nav>
</div>
<div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
</div>
<main class="content quarto-banner-title-block" id="quarto-document-content">





<section id="introduction" class="level2" data-number="1">
<h2 data-number="1" class="anchored" data-anchor-id="introduction"><span class="header-section-number">1</span> Introduction</h2>
<p>Most phenotypes, from organism to cell, emerge from complex interactions of biological elements. Reductionist approaches to the elucidation of biological processes have gained immense popularity due to the relative simplicity and clear nature of such research workflows, focusing on a single molecule or a few of them at a time. However, in order to understand life from its simplest to its most complex forms, we need to delve into the world of networks.</p>
<p>Network theory has long been studied because of the many fields it can be applied to. The majority of complex systems can be modeled or explained by networks with, more or less, the same architectural features, highlighting the universality of such systems and the laws they are governed by. Despite disparities in the details that characterize each network, the abstract nature of network theory and analyses allow for the extraction of significant information that can help with comprehending critical constituents of a network’s organization, function and robustness, and can have major impacts on the fields of interest and their respective advancements <span class="citation" data-cites="barabási2004">(<a href="#ref-barabási2004" role="doc-biblioref">Barabási and Oltvai 2004</a>)</span>.</p>
<p>The following protein network analysis is performed on protein-protein interaction (PPI) data that were derived from two-hybrid experiments. These data constitute part of a differential epistasis mapping paper published by Bandyopadhyay et. al.&nbsp;in 2010 <span class="citation" data-cites="bandyopadhyay2010">(<a href="#ref-bandyopadhyay2010" role="doc-biblioref">Bandyopadhyay et al. 2010</a>)</span>. Pertaining to comparison of genetic networks across conditions, this study reveals differences in interactions of DNA damage response genes with and without cellular exposure to methane methylsulfonate or MMS, a DNA-damaging agent.</p>
<p>Two-hybrid experiments are a set of molecular biology techniques that aim to study protein-protein interactions and allow researchers to determine if two proteins can physically interact with each other in a living cell. The most common type of two-hybrid system is the yeast two-hybrid (Y2H) system, although variations exist for use in other organisms.</p>
<p>The process consists of a bait protein of interest that is fused to a DNA-binding domain (DBD) and a prey, which is another protein of interest, fused to an activation domain (AD). The system’s function relies on the reconstitution of a transcription factor when the two proteins of interest interact inside the living cell. If such interaction does occur after the cellular introduction of the bait and prey proteins, the functional transcription factor will bind to the promoter of a reporter gene. Through this process, protein-protein interactions can be detected in vivo and lead to advancements in the study of such complex networks that underlie cellular processes <span class="citation" data-cites="suter2008">(<a href="#ref-suter2008" role="doc-biblioref">Suter, Kittanakom, and Stagljar 2008</a>)</span>.</p>
<p>This brief report is part of a personal project exploring the world of complex systems, network biology and the statistic measures that describe their properties and behavior. The aim of this project is to better understand biological interactions and network theory, while getting familiar with using Quarto publishing system and the appropriate R libraries for graph visualization, as well as the calculation of a few selected network metrics, including node degree and degree distribution, betweenness centrality, transitivity, modularity and community detection.</p>
</section>
<section id="results" class="level2" data-number="2">
<h2 data-number="2" class="anchored" data-anchor-id="results"><span class="header-section-number">2</span> Results</h2>
<section id="graph-visualization" class="level3" data-number="2.1">
<h3 data-number="2.1" class="anchored" data-anchor-id="graph-visualization"><span class="header-section-number">2.1</span> Graph Visualization</h3>
<p>A network graph is created using only the unique nodes of the protein network and the appropriate edges between them. In this case, the generated graph is undirected, meaning its edges do not point to a defined direction, from or to a specific node, in each path of the network.</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a>knitr<span class="sc">::</span>opts_chunk<span class="sc">$</span><span class="fu">set</span>(<span class="at">message =</span> <span class="cn">FALSE</span>, <span class="at">warning =</span> <span class="cn">FALSE</span>)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">options</span>(<span class="at">warn =</span> <span class="sc">-</span><span class="dv">1</span>)</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="co">#install.packages("ggraph")</span></span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="co">#install.packages("tidygraph")</span></span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="co">#install.packages("ggplot2")</span></span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="co">#install.packages("igraph",  repos = "https://cran.r-project.org")</span></span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a><span class="co">#install.packages("gt")</span></span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a><span class="fu">suppressPackageStartupMessages</span>({</span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">library</span>(igraph)</span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">library</span>(ggraph)</span>
<span id="cb1-12"><a href="#cb1-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">library</span>(tidygraph)</span>
<span id="cb1-13"><a href="#cb1-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">library</span>(gt)</span>
<span id="cb1-14"><a href="#cb1-14" aria-hidden="true" tabindex="-1"></a>})</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Read data from txt file</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>data <span class="ot">&lt;-</span> <span class="fu">read.table</span>(<span class="st">"Bandyopadhyay2010.txt"</span>, <span class="at">header =</span> T)</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Isolate only unique proteins of dataset</span></span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a>unique_nodes <span class="ot">&lt;-</span> <span class="fu">unique</span>(<span class="fu">c</span>(data<span class="sc">$</span>BAIT_OFFICIAL_SYMBOL, data<span class="sc">$</span>PREY_OFFICIAL_SYMBOL))</span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Create edges by binding together columns of initial dataset</span></span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a>edges <span class="ot">&lt;-</span> <span class="fu">cbind</span>(data<span class="sc">$</span>BAIT_OFFICIAL_SYMBOL, data<span class="sc">$</span>PREY_OFFICIAL_SYMBOL)</span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-10"><a href="#cb2-10" aria-hidden="true" tabindex="-1"></a><span class="co"># Create graph object</span></span>
<span id="cb2-11"><a href="#cb2-11" aria-hidden="true" tabindex="-1"></a>graph <span class="ot">&lt;-</span> <span class="fu">graph_from_data_frame</span>(edges, <span class="at">directed =</span> F)</span>
<span id="cb2-12"><a href="#cb2-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-13"><a href="#cb2-13" aria-hidden="true" tabindex="-1"></a><span class="co"># Add label attribute to graph vertices for protein names</span></span>
<span id="cb2-14"><a href="#cb2-14" aria-hidden="true" tabindex="-1"></a><span class="fu">V</span>(graph)<span class="sc">$</span>label <span class="ot">&lt;-</span> unique_nodes</span>
<span id="cb2-15"><a href="#cb2-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-16"><a href="#cb2-16" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot graph with node size according to node degrees</span></span>
<span id="cb2-17"><a href="#cb2-17" aria-hidden="true" tabindex="-1"></a><span class="fu">ggraph</span>(graph, <span class="at">layout =</span> <span class="st">"fr"</span>) <span class="sc">+</span> </span>
<span id="cb2-18"><a href="#cb2-18" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_edge_link</span>(<span class="at">alpha =</span> <span class="fl">0.25</span>) <span class="sc">+</span> </span>
<span id="cb2-19"><a href="#cb2-19" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_node_point</span>(<span class="fu">aes</span>(<span class="at">size =</span> <span class="fu">degree</span>(graph)), <span class="at">shape =</span> <span class="dv">21</span>, <span class="at">fill =</span> <span class="st">"#7ea695"</span>, <span class="at">color =</span><span class="st">"black"</span>) <span class="sc">+</span> </span>
<span id="cb2-20"><a href="#cb2-20" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_void</span>() <span class="sc">+</span></span>
<span id="cb2-21"><a href="#cb2-21" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">"Protein-Protein Interaction Network"</span>) <span class="sc">+</span></span>
<span id="cb2-22"><a href="#cb2-22" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">plot.title =</span> <span class="fu">element_text</span>(<span class="at">hjust =</span> <span class="fl">0.5</span>, <span class="at">margin =</span> <span class="fu">margin</span>(<span class="at">t =</span><span class="dv">20</span>)),</span>
<span id="cb2-23"><a href="#cb2-23" aria-hidden="true" tabindex="-1"></a>        <span class="at">plot.title.position =</span> <span class="st">"plot"</span>,</span>
<span id="cb2-24"><a href="#cb2-24" aria-hidden="true" tabindex="-1"></a>        <span class="at">plot.margin =</span> <span class="fu">margin</span>(<span class="at">t =</span> <span class="dv">30</span>, <span class="at">r =</span> <span class="dv">10</span>, <span class="at">b =</span> <span class="dv">10</span>, <span class="at">l =</span> <span class="dv">10</span>))<span class="sc">+</span></span>
<span id="cb2-25"><a href="#cb2-25" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">size =</span> <span class="st">"Node Degree"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div id="fig-graph" class="quarto-figure quarto-figure-center quarto-float anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-graph-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="BioNet_files/figure-html/fig-graph-1.png" class="img-fluid figure-img" width="672">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-graph-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;1: Graph of protein-protein interaction network with node size being proportional to node degrees.
</figcaption>
</figure>
</div>
</div>
</div>
<p>Plotting the graph as it is leads to an output that has too much information to be interpreted and, thus, further analysis is needed in order to get to meaningful conclusions. It does, however, convey a few propensities characteristic of biological networks, such as the high degree of clustering, with the obvious existence of closely linked groups in the system around central nodes, so-called “hubs”. Also, the graph indicates the degree of different nodes, with a few nodes having a great number of links and forming big groups and most of the vertices being connected to little to no other nodes, hinting to the scale-free nature of the network. This is a tendency especially prominent in PPI and protein domain networks that denotes the intertwined modularity of such systems; proteins mostly work in functional and structural groups that cooperate to give rise to cellular phenotypes <span class="citation" data-cites="barabási2004">(<a href="#ref-barabási2004" role="doc-biblioref">Barabási and Oltvai 2004</a>)</span>.</p>
</section>
<section id="node-degree-k" class="level3" data-number="2.2">
<h3 data-number="2.2" class="anchored" data-anchor-id="node-degree-k"><span class="header-section-number">2.2</span> Node Degree (k)</h3>
<p>In the context of network theory and graph analysis, node degree refers to the number of edges or direct connections to other vertices that are incident to a specific node in the graph <span class="citation" data-cites="opsahl2010">(<a href="#ref-opsahl2010" role="doc-biblioref">Opsahl, Agneessens, and Skvoretz 2010</a>)</span>. High-degree nodes, often referred to as “hubs”, are typically crucial for the network’s structure and function and they often represent key proteins or genes in biological networks. On the contrary, low-degree nodes might pertain to more specialized or peripheral components in the system, since they have less interactions with the rest of the vertices. Understanding the information that is denoted by node degree significantly helps in analyzing the importance and influence of specific nodes in the network, elucidating the overall network topology and even identifying critical points for potential intervention <span class="citation" data-cites="przulj2004">(<a href="#ref-przulj2004" role="doc-biblioref">Pržulj, Corneil, and Jurisica 2004</a>)</span>.</p>
<section id="degree-distribution-pk" class="level4" data-number="2.2.1">
<h4 data-number="2.2.1" class="anchored" data-anchor-id="degree-distribution-pk"><span class="header-section-number">2.2.1</span> Degree distribution (P(k))</h4>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate node degrees of graph</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>node_degrees <span class="ot">&lt;-</span> <span class="fu">degree</span>(graph)</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Create frequency table of node degrees</span></span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a>degree_count <span class="ot">&lt;-</span> <span class="fu">table</span>(node_degrees)</span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate degree distribution of graph nodes</span></span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true" tabindex="-1"></a>deg_dist <span class="ot">&lt;-</span> degree_count <span class="sc">/</span> <span class="fu">vcount</span>(graph)</span>
<span id="cb3-9"><a href="#cb3-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-10"><a href="#cb3-10" aria-hidden="true" tabindex="-1"></a><span class="co"># Turn into dataframe</span></span>
<span id="cb3-11"><a href="#cb3-11" aria-hidden="true" tabindex="-1"></a>deg_df <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">Degree =</span> <span class="fu">as.numeric</span>(<span class="fu">names</span>(deg_dist)), <span class="at">Frequency =</span> <span class="fu">as.numeric</span>(deg_dist))</span>
<span id="cb3-12"><a href="#cb3-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-13"><a href="#cb3-13" aria-hidden="true" tabindex="-1"></a><span class="co"># Degree distribution barplot</span></span>
<span id="cb3-14"><a href="#cb3-14" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(deg_df, <span class="fu">aes</span>(<span class="at">x =</span> Degree, <span class="at">y =</span> Frequency)) <span class="sc">+</span> </span>
<span id="cb3-15"><a href="#cb3-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_bar</span>(<span class="at">stat =</span> <span class="st">"identity"</span>, <span class="at">fill =</span> <span class="st">"#7ea695"</span>,<span class="at">color =</span> <span class="st">"black"</span>) <span class="sc">+</span></span>
<span id="cb3-16"><a href="#cb3-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"Degree Distribution (P(k))"</span>, <span class="at">x =</span> <span class="st">"Degree (k)"</span>, <span class="at">y =</span> <span class="st">"Frequency"</span>) <span class="sc">+</span></span>
<span id="cb3-17"><a href="#cb3-17" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_minimal</span>() <span class="sc">+</span></span>
<span id="cb3-18"><a href="#cb3-18" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">plot.title =</span> <span class="fu">element_text</span>(<span class="at">hjust =</span> <span class="fl">0.5</span>, <span class="at">margin =</span> <span class="fu">margin</span>(<span class="at">t=</span><span class="dv">20</span>)))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div id="fig-degree-distribution" class="quarto-figure quarto-figure-center quarto-float anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-degree-distribution-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="BioNet_files/figure-html/fig-degree-distribution-1.png" class="img-fluid figure-img" width="672">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-degree-distribution-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;2: Barplot of degree distribution of nodes in PPI network.
</figcaption>
</figure>
</div>
</div>
</div>
<p>Degree distribution (P(k)) is a network property that gives the probability for a randomly chosen node to have k links or, in other words, a degree equal to k. It is defined as follows:</p>
<p><span class="math display">
P(k) = \frac{\left|\{i \mid \deg(i) = k\}\right|}{n}
</span></p>
<p>where the numerator represents the number of nodes in the graph that have degree equal to <span class="math inline">{\displaystyle k}</span> and the denominator <span class="math inline">{\displaystyle n}</span> is the total number of nodes in the network.</p>
<p>Degree distribution conveys information about how the system is structured and allows for distinguishing between different network classes. Degree distribution of random networks tends to follow the Poisson distribution, with most of the nodes having the same amount of edges or links, a value that is closer to the average <span class="math inline">{\displaystyle &lt;k&gt;}</span>. On the contrary, real complex systems, including PPI networks, showcase a highly asymmetric degree distribution that follows Power law and has a right skewness, which is obvious in <a href="#fig-degree-distribution" class="quarto-xref">Figure&nbsp;2</a>. It clearly denotes that the vast majority of the system’s nodes have a low degree, close or equal to 1, while a few vertices constitute hubs with extraordinarily high values of degrees. This is a characteristic attribute of scale-free networks, which mostly describe complex systems of the real world. Briefly, this property indicates the absence of a typical element or node in the network that could describe all other nodes, which could exist in random networks, where most node degrees approximate the average degree, but not in real networks <span class="citation" data-cites="barabási2004">(<a href="#ref-barabási2004" role="doc-biblioref">Barabási and Oltvai 2004</a>)</span>.</p>
<p>Node degree and degree distribution have significant consequences for the network’s function and robustness. Generally, robustness is a property that pertains to the system’s capacity to react to changes in its environment or internal structure, while continuing to behave in a relatively typical manner. As its immunology counterpart of homeostasis, robustness is critical to the viability of the network, especially since most complex systems in the real world interact with highly variable and dynamic environments <span class="citation" data-cites="albert2000">(<a href="#ref-albert2000" role="doc-biblioref">Albert, Jeong, and Barabási 2000</a>)</span>.</p>
<p>There is a strong relationship between the hub status of a node and its role in maintaining the network’s connectivity and viability, especially when it comes to topological robustness. Generally, scale-free networks with long tail degree distributions are resilient to random removal of nodes, a phenomenon known as “failure”, but they are vulnerable to the removal of hub vertices, which is known as “attack”. Thus, this type of networks are robust against accidental failures, since most of their components are low degree nodes, they do, however, rely on hubs for their structure and flow of information. As a result, the attack vulnerability of such networks can potentially affect both dynamical and functional robustness, besides the obviously afflicted topological one <span class="citation" data-cites="liu2017">(<a href="#ref-liu2017" role="doc-biblioref">Liu et al. 2017</a>)</span>.</p>
</section>
</section>
<section id="clustering-and-communities" class="level3" data-number="2.3">
<h3 data-number="2.3" class="anchored" data-anchor-id="clustering-and-communities"><span class="header-section-number">2.3</span> Clustering and Communities</h3>
<section id="community-detection" class="level4" data-number="2.3.1">
<h4 data-number="2.3.1" class="anchored" data-anchor-id="community-detection"><span class="header-section-number">2.3.1</span> Community Detection</h4>
<p>For better visualization of protein-protein interactions, the vertices that pertain to unique proteins of the given network can be grouped into clusters that constitute protein communities; that is, groups of proteins that seem to interact with each other in a greater degree than they do with other proteins of the network. Here, community detection was performed using the Louvain clustering algorithm, a popular modularity-based algorithm for the identification of communities in large networks. This method is widely preferred due to its agglomerative nature, since it is based on a multi-scale modularity optimization process. However, it must be pointed out that the Louvain algorithm does have a resolution limit that stems from its size-dependence; the larger the system, the higher the probability for detection of larger clusters/communities <span class="citation" data-cites="blondel2008">(<a href="#ref-blondel2008" role="doc-biblioref">Blondel et al. 2008</a>)</span>.</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Apply Louvain clustering algorithm to graph for community detection</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>net_communities <span class="ot">&lt;-</span> <span class="fu">cluster_louvain</span>(graph)</span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Extract community membership vector from community detection result object</span></span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Indicates which community each node in the network belongs to</span></span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a>membership <span class="ot">&lt;-</span> <span class="fu">membership</span>(net_communities)</span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-8"><a href="#cb4-8" aria-hidden="true" tabindex="-1"></a><span class="co"># Convert graph object into tbl_graph object for better visualization of graph</span></span>
<span id="cb4-9"><a href="#cb4-9" aria-hidden="true" tabindex="-1"></a>tbl_graph <span class="ot">&lt;-</span> <span class="fu">as_tbl_graph</span>(graph)</span>
<span id="cb4-10"><a href="#cb4-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-11"><a href="#cb4-11" aria-hidden="true" tabindex="-1"></a><span class="co"># Add community attribute to tbl_graph vertices to match each node to its community</span></span>
<span id="cb4-12"><a href="#cb4-12" aria-hidden="true" tabindex="-1"></a><span class="fu">V</span>(tbl_graph)<span class="sc">$</span>community <span class="ot">&lt;-</span> membership</span>
<span id="cb4-13"><a href="#cb4-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-14"><a href="#cb4-14" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot graph with node colors according to community membership of each node</span></span>
<span id="cb4-15"><a href="#cb4-15" aria-hidden="true" tabindex="-1"></a><span class="co"># using the Fruchterman-Reingold layout </span></span>
<span id="cb4-16"><a href="#cb4-16" aria-hidden="true" tabindex="-1"></a><span class="fu">ggraph</span>(tbl_graph, <span class="at">layout =</span> <span class="st">"fr"</span>) <span class="sc">+</span></span>
<span id="cb4-17"><a href="#cb4-17" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_edge_link</span>(<span class="at">alpha =</span> <span class="fl">0.25</span>) <span class="sc">+</span> </span>
<span id="cb4-18"><a href="#cb4-18" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_node_point</span>(<span class="fu">aes</span>(<span class="at">fill =</span> <span class="fu">as.factor</span>(community)), <span class="at">size =</span> <span class="dv">2</span>,</span>
<span id="cb4-19"><a href="#cb4-19" aria-hidden="true" tabindex="-1"></a>                  <span class="at">shape =</span> <span class="dv">21</span>, <span class="at">stroke =</span> <span class="fl">0.08</span>, <span class="at">border =</span> <span class="st">"grey"</span>) <span class="sc">+</span></span>
<span id="cb4-20"><a href="#cb4-20" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_void</span>() <span class="sc">+</span></span>
<span id="cb4-21"><a href="#cb4-21" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"PPI Network Communities"</span>,</span>
<span id="cb4-22"><a href="#cb4-22" aria-hidden="true" tabindex="-1"></a>       <span class="at">subtitle =</span> <span class="st">"Louvain Clustering"</span>,</span>
<span id="cb4-23"><a href="#cb4-23" aria-hidden="true" tabindex="-1"></a>       <span class="at">fill =</span> <span class="st">"Community"</span>) <span class="sc">+</span></span>
<span id="cb4-24"><a href="#cb4-24" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">plot.title =</span> <span class="fu">element_text</span>(<span class="at">hjust =</span> <span class="fl">0.5</span>),</span>
<span id="cb4-25"><a href="#cb4-25" aria-hidden="true" tabindex="-1"></a>        <span class="at">plot.subtitle =</span> <span class="fu">element_text</span>(<span class="at">hjust =</span> <span class="fl">0.5</span>, <span class="at">vjust =</span> <span class="sc">-</span><span class="dv">1</span>),</span>
<span id="cb4-26"><a href="#cb4-26" aria-hidden="true" tabindex="-1"></a>        <span class="at">legend.key.height =</span> <span class="fu">unit</span>(<span class="fl">0.1</span>, <span class="st">"line"</span>))<span class="sc">+</span></span>
<span id="cb4-27"><a href="#cb4-27" aria-hidden="true" tabindex="-1"></a>  <span class="fu">guides</span>(<span class="at">fill =</span> <span class="fu">guide_legend</span>(<span class="at">ncol =</span> <span class="dv">2</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div id="fig-louvain-communities" class="quarto-figure quarto-figure-center quarto-float anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-louvain-communities-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="BioNet_files/figure-html/fig-louvain-communities-1.png" class="img-fluid figure-img" width="672">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-louvain-communities-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;3: Graph of PPI network after applying the Louvain clustering algorithm for community detection. Communities are discerned by color.
</figcaption>
</figure>
</div>
</div>
</div>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Assign membership vector to variable com_member_net</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>com_member_net <span class="ot">&lt;-</span> membership</span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Identify all unique communities in the network</span></span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a>unique_com_net <span class="ot">&lt;-</span> <span class="fu">unique</span>(membership)</span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Create placeholder list variable</span></span>
<span id="cb5-8"><a href="#cb5-8" aria-hidden="true" tabindex="-1"></a>community_sizes_net <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb5-9"><a href="#cb5-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-10"><a href="#cb5-10" aria-hidden="true" tabindex="-1"></a><span class="co"># Iterate through unique communities of network to count each community's nodes</span></span>
<span id="cb5-11"><a href="#cb5-11" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(unique_com_net)) {</span>
<span id="cb5-12"><a href="#cb5-12" aria-hidden="true" tabindex="-1"></a>  nodes_in_com <span class="ot">&lt;-</span> <span class="fu">which</span>(com_member_net <span class="sc">==</span> unique_com_net[i])</span>
<span id="cb5-13"><a href="#cb5-13" aria-hidden="true" tabindex="-1"></a>  community_sizes_net[[i]] <span class="ot">&lt;-</span> <span class="fu">length</span>(nodes_in_com)</span>
<span id="cb5-14"><a href="#cb5-14" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb5-15"><a href="#cb5-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-16"><a href="#cb5-16" aria-hidden="true" tabindex="-1"></a><span class="co"># Create dataframe of community sizes</span></span>
<span id="cb5-17"><a href="#cb5-17" aria-hidden="true" tabindex="-1"></a>com_size_net_df <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">Community_ID =</span> <span class="fu">as.factor</span>(unique_com_net),</span>
<span id="cb5-18"><a href="#cb5-18" aria-hidden="true" tabindex="-1"></a>                              <span class="at">Size =</span> <span class="fu">as.numeric</span>(<span class="fu">unlist</span>(community_sizes_net)))</span>
<span id="cb5-19"><a href="#cb5-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-20"><a href="#cb5-20" aria-hidden="true" tabindex="-1"></a><span class="co"># Barplot of community sizes</span></span>
<span id="cb5-21"><a href="#cb5-21" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(com_size_net_df, <span class="fu">aes</span>(<span class="at">x=</span> Community_ID, <span class="at">y=</span> Size, <span class="at">fill=</span> Size)) <span class="sc">+</span></span>
<span id="cb5-22"><a href="#cb5-22" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_bar</span>(<span class="at">stat =</span> <span class="st">"identity"</span>)<span class="sc">+</span></span>
<span id="cb5-23"><a href="#cb5-23" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"Community Sizes"</span>, <span class="at">x =</span> <span class="st">"Community"</span>, <span class="at">y =</span> <span class="st">"Number Of Nodes"</span>)<span class="sc">+</span></span>
<span id="cb5-24"><a href="#cb5-24" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_minimal</span>() <span class="sc">+</span></span>
<span id="cb5-25"><a href="#cb5-25" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">plot.title.position=</span> <span class="st">"plot"</span>,</span>
<span id="cb5-26"><a href="#cb5-26" aria-hidden="true" tabindex="-1"></a>        <span class="at">plot.title =</span> <span class="fu">element_text</span>(<span class="at">hjust =</span> <span class="fl">0.5</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div id="fig-community-sizes" class="quarto-figure quarto-figure-center quarto-float anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-community-sizes-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="BioNet_files/figure-html/fig-community-sizes-1.png" class="img-fluid figure-img" width="672">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-community-sizes-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;4: Barplot of community sizes in PPI network.
</figcaption>
</figure>
</div>
</div>
</div>
<p>After applying the Louvain clustering algorithm to the network for community detection, it is clear that there are 21 communities, each one indicated by a different color in <a href="#fig-louvain-communities" class="quarto-xref">Figure&nbsp;3</a>. <a href="#fig-community-sizes" class="quarto-xref">Figure&nbsp;4</a> shows the sizes of the communities detected in the network, with over half of the communities having more than 20 nodes as members, and 4 of them having 50 or more members.</p>
</section>
<section id="modularity" class="level4" data-number="2.3.2">
<h4 data-number="2.3.2" class="anchored" data-anchor-id="modularity"><span class="header-section-number">2.3.2</span> Modularity</h4>
<p>Modularity measures the extend to which the network is organized into communities. The modularity of a given set of communities within a network is defined as follows:</p>
<p><span class="math display">
Q = \frac{|E_{in}| - \langle E_{in} \rangle}{|E|}
</span></p>
<p>where <span class="math inline">|Ein|</span> is the number of edges within a community that do not cross boundaries between communities, <span class="math inline">\langle Ein \rangle</span> is the expected number of within-community edges if the topology were completely random and <span class="math inline">|E|</span> is the total number of edges. The value of modularity for a specific network ranges from -1 to 1, with positive values denoting that the number of edges within communities are more than the expected number.</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Modularity calculates the modularity of a graph with respect to the given membership vector</span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>modularity_score <span class="ot">&lt;-</span> <span class="fu">modularity</span>(graph, membership, <span class="at">resolution =</span> <span class="dv">1</span>)</span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a>rounded_modularity <span class="ot">&lt;-</span> <span class="fu">round</span>(modularity_score, <span class="dv">2</span>)</span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">"Q ="</span>, rounded_modularity)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>Q = 0.75</code></pre>
</div>
</div>
<p>The network in question has a modularity score of 0.75, which indicates a relatively strong community structure, with well-clustered nodes, many edges within communities and few edges between discrete communities. In biological terms, modularity refers to groups of interlinked molecules, such as proteins, that cooperate to achieve a specific function. As expected, most complex biological systems, like PPI networks, usually brim with modular functionality and/or structure, a property that is correlated with high network robustness <span class="citation" data-cites="newman2006">(<a href="#ref-newman2006" role="doc-biblioref">Newman 2006</a>)</span>.</p>
</section>
<section id="betweenness-centrality" class="level4" data-number="2.3.3">
<h4 data-number="2.3.3" class="anchored" data-anchor-id="betweenness-centrality"><span class="header-section-number">2.3.3</span> Betweenness Centrality</h4>
<p>Betweenness centrality denotes the number of times a node acts as a bridge along the shortest path between two other nodes. The betweenness centrality of a node <span class="math inline">{\displaystyle v}</span> is defined as follows:</p>
<p><span class="math display">
{\displaystyle g(v)=\sum _{s\neq v\neq t}{\frac {\sigma _{st}(v)}{\sigma _{st}}}}
</span></p>
<p>where <span class="math inline">{\displaystyle \sigma _{st}}</span> is the number of shortest paths from node <span class="math inline">{\displaystyle s}</span> to node <span class="math inline">{\displaystyle t}</span> and <span class="math inline">{\displaystyle \sigma _{st}(v)}</span> is the number of shortest paths from node <span class="math inline">{\displaystyle s}</span> to node <span class="math inline">{\displaystyle t}</span> that go through node <span class="math inline">{\displaystyle v}</span> (without node <span class="math inline">{\displaystyle v}</span> being an ending or starting point).</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate betweenness centrality of each node in graph</span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>betweenness_values <span class="ot">&lt;-</span> <span class="fu">betweenness</span>(graph, <span class="at">normalized =</span> T)</span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Add attribute Betweenness to tbl_graph vertices</span></span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a><span class="fu">V</span>(tbl_graph)<span class="sc">$</span>Betweenness <span class="ot">&lt;-</span> betweenness_values</span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Create vector of only non zero betweenness centrality values</span></span>
<span id="cb8-8"><a href="#cb8-8" aria-hidden="true" tabindex="-1"></a>non_zero_betweeness <span class="ot">&lt;-</span> betweenness_values[betweenness_values <span class="sc">&gt;</span> <span class="dv">0</span>]</span>
<span id="cb8-9"><a href="#cb8-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-10"><a href="#cb8-10" aria-hidden="true" tabindex="-1"></a><span class="co"># Create vector of only nodes with non-zero betweenness centrality values</span></span>
<span id="cb8-11"><a href="#cb8-11" aria-hidden="true" tabindex="-1"></a>non_zero_names <span class="ot">&lt;-</span> <span class="fu">V</span>(tbl_graph)<span class="sc">$</span>name[(<span class="fu">V</span>(tbl_graph)<span class="sc">$</span>Betweenness) <span class="sc">&gt;</span> <span class="dv">0</span>]</span>
<span id="cb8-12"><a href="#cb8-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-13"><a href="#cb8-13" aria-hidden="true" tabindex="-1"></a><span class="co"># Create dataframe of nodes and non-zero betweenness centrality values</span></span>
<span id="cb8-14"><a href="#cb8-14" aria-hidden="true" tabindex="-1"></a>betweenness_df <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">Proteins =</span> <span class="fu">as.factor</span>(non_zero_names),</span>
<span id="cb8-15"><a href="#cb8-15" aria-hidden="true" tabindex="-1"></a>                             <span class="at">Betweenness =</span> <span class="fu">as.numeric</span>(non_zero_betweeness))</span>
<span id="cb8-16"><a href="#cb8-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-17"><a href="#cb8-17" aria-hidden="true" tabindex="-1"></a><span class="co"># Betweenness centrality barplot</span></span>
<span id="cb8-18"><a href="#cb8-18" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(betweenness_df, <span class="fu">aes</span>(<span class="at">x =</span>Proteins, <span class="at">y =</span> Betweenness))<span class="sc">+</span></span>
<span id="cb8-19"><a href="#cb8-19" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_bar</span>(<span class="at">stat =</span> <span class="st">"identity"</span>)<span class="sc">+</span></span>
<span id="cb8-20"><a href="#cb8-20" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_minimal</span>()<span class="sc">+</span></span>
<span id="cb8-21"><a href="#cb8-21" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"Betweenness Centrality of Proteins"</span>,</span>
<span id="cb8-22"><a href="#cb8-22" aria-hidden="true" tabindex="-1"></a>       <span class="at">x =</span> <span class="st">"Nodes"</span>,</span>
<span id="cb8-23"><a href="#cb8-23" aria-hidden="true" tabindex="-1"></a>       <span class="at">y =</span> <span class="st">"Betweenness Centrality"</span>) <span class="sc">+</span></span>
<span id="cb8-24"><a href="#cb8-24" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">plot.title =</span> <span class="fu">element_text</span>(<span class="at">hjust=</span><span class="fl">0.5</span>),</span>
<span id="cb8-25"><a href="#cb8-25" aria-hidden="true" tabindex="-1"></a>        <span class="at">axis.text.x =</span> <span class="fu">element_blank</span>())</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div id="fig-betweenness-centrality" class="quarto-figure quarto-figure-center quarto-float anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-betweenness-centrality-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="BioNet_files/figure-html/fig-betweenness-centrality-1.png" class="img-fluid figure-img" width="672">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-betweenness-centrality-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;5: Barplot of betweenness centrality values of nodes in PPI network.
</figcaption>
</figure>
</div>
</div>
</div>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Create vector of only relatively high betweenness centrality values</span></span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Betweenness centrality &gt; 0.1</span></span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a>high_betweenness <span class="ot">&lt;-</span> betweenness_values[betweenness_values <span class="sc">&gt;</span> <span class="fl">0.1</span>]</span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Create vector of nodes with betweenness centrality &gt; 0.1</span></span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true" tabindex="-1"></a>high_betweenness_nodes <span class="ot">&lt;-</span> <span class="fu">V</span>(tbl_graph)<span class="sc">$</span>name[(<span class="fu">V</span>(tbl_graph)<span class="sc">$</span>Betweenness) <span class="sc">&gt;</span> <span class="fl">0.1</span>]</span>
<span id="cb9-7"><a href="#cb9-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-8"><a href="#cb9-8" aria-hidden="true" tabindex="-1"></a><span class="co"># Create dataframe of nodes and betweenness centrality values [&gt; 0.1]</span></span>
<span id="cb9-9"><a href="#cb9-9" aria-hidden="true" tabindex="-1"></a>high_betweenness_df <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">Proteins =</span> <span class="fu">as.factor</span>(high_betweenness_nodes),</span>
<span id="cb9-10"><a href="#cb9-10" aria-hidden="true" tabindex="-1"></a>                             <span class="at">Betweenness =</span> <span class="fu">as.numeric</span>(high_betweenness))</span>
<span id="cb9-11"><a href="#cb9-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-12"><a href="#cb9-12" aria-hidden="true" tabindex="-1"></a><span class="co"># Barplot of nodes with betweenness centrality &gt; 0.1</span></span>
<span id="cb9-13"><a href="#cb9-13" aria-hidden="true" tabindex="-1"></a><span class="co"># Central nodes for information flow in network</span></span>
<span id="cb9-14"><a href="#cb9-14" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(high_betweenness_df, <span class="fu">aes</span>(<span class="at">x =</span> Proteins, <span class="at">y =</span> Betweenness))<span class="sc">+</span></span>
<span id="cb9-15"><a href="#cb9-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_bar</span>(<span class="at">stat =</span> <span class="st">"identity"</span>, <span class="at">fill =</span> <span class="st">"#7ea695"</span>, <span class="at">color =</span> <span class="st">"black"</span>, <span class="at">width=</span> <span class="fl">0.8</span>)<span class="sc">+</span></span>
<span id="cb9-16"><a href="#cb9-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_minimal</span>()<span class="sc">+</span></span>
<span id="cb9-17"><a href="#cb9-17" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"Proteins with Betweenness Centrality &gt; 0.1"</span>,</span>
<span id="cb9-18"><a href="#cb9-18" aria-hidden="true" tabindex="-1"></a>       <span class="at">x =</span> <span class="st">"Nodes"</span>,</span>
<span id="cb9-19"><a href="#cb9-19" aria-hidden="true" tabindex="-1"></a>       <span class="at">y =</span> <span class="st">"Betweenness Centrality"</span>) <span class="sc">+</span></span>
<span id="cb9-20"><a href="#cb9-20" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">plot.title =</span> <span class="fu">element_text</span>(<span class="at">hjust=</span><span class="fl">0.5</span>),</span>
<span id="cb9-21"><a href="#cb9-21" aria-hidden="true" tabindex="-1"></a>        <span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">angle=</span><span class="dv">90</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div id="fig-high-betweenness-nodes" class="quarto-figure quarto-figure-center quarto-float anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-high-betweenness-nodes-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="BioNet_files/figure-html/fig-high-betweenness-nodes-1.png" class="img-fluid figure-img" width="672">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-high-betweenness-nodes-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;6: Barplot of nodes with betweenness centrality values of &gt;0.1.
</figcaption>
</figure>
</div>
</div>
</div>
<p>As shown in <a href="#fig-betweenness-centrality" class="quarto-xref">Figure&nbsp;5</a>, most of the network’s nodes have low betweenness centrality values, with only a few exceeding 0.1. The proteins with relatively high betweenness centrality values (&gt;0.1) are plotted in <a href="#fig-high-betweenness-nodes" class="quarto-xref">Figure&nbsp;6</a>, with APC or adenomatous polyposis coli, a Wnt signalling pathway regulator, reaching over 0.3. High betweenness centrality indicates that a node significantly influences the flow of information within a network and is crucial for maintaining the network’s functionality. Usually, such proteins are evolutionarily conserved and constitute some of the oldest vertices in the PPI system, as is APC, which exhibits an extreme degree of conservation across taxa with most mutations having adverse effects for cellular processes, especially in the epithelium <span class="citation" data-cites="abbott2023">(<a href="#ref-abbott2023" role="doc-biblioref">Abbott and Näthke 2023</a>)</span>.</p>
</section>
<section id="transitivity" class="level4" data-number="2.3.4">
<h4 data-number="2.3.4" class="anchored" data-anchor-id="transitivity"><span class="header-section-number">2.3.4</span> Transitivity</h4>
<p>Transitivity, also known as global clustering coefficient, quantifies the propensity of a group of nodes to cluster together in a network. It is defined as the ratio of the number of closed triplets or triangles, to the number of all triplets, closed and open, of the network. Global clustering coefficient provides an average measure of the clustering in a network and is defined as follows:</p>
<p><span class="math display">
C_T = \frac{\left| \{ (i, j, k) \mid d(i, j) = d(i, k) = d(j, k) = 1 \} \right|}{\left| \{ (i, j, k) \mid d(i, j) = d(i, k) = 1 \} \right|}
</span></p>
<p>where the denominator is the total number of possible node pairs within node <span class="math inline">{\displaystyle i}</span> ’s neighborhood and the numerator is the number of actually connected node pairs among them.</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate transitivity values of nodes in network</span></span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>transitivity_values <span class="ot">&lt;-</span> <span class="fu">transitivity</span>(graph, <span class="at">type =</span> <span class="st">"local"</span>)</span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Add transitivity values as attribute of tbl_graph vertices</span></span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a><span class="fu">V</span>(tbl_graph)<span class="sc">$</span>Transitivity <span class="ot">&lt;-</span> transitivity_values</span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-7"><a href="#cb10-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Create vector of only non-zero transitivity values</span></span>
<span id="cb10-8"><a href="#cb10-8" aria-hidden="true" tabindex="-1"></a>non_zero_trans <span class="ot">&lt;-</span> transitivity_values[transitivity_values <span class="sc">&gt;</span> <span class="dv">0</span>]</span>
<span id="cb10-9"><a href="#cb10-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-10"><a href="#cb10-10" aria-hidden="true" tabindex="-1"></a><span class="co"># Create vector of nodes with non-zero transitivity values</span></span>
<span id="cb10-11"><a href="#cb10-11" aria-hidden="true" tabindex="-1"></a>non_zero_trans_nodes <span class="ot">&lt;-</span> <span class="fu">V</span>(tbl_graph)<span class="sc">$</span>name[(<span class="fu">V</span>(tbl_graph)<span class="sc">$</span>Transitivity) <span class="sc">&gt;</span> <span class="dv">0</span>]</span>
<span id="cb10-12"><a href="#cb10-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-13"><a href="#cb10-13" aria-hidden="true" tabindex="-1"></a><span class="co"># Create dataframe of nodes and non-zero transitivity values</span></span>
<span id="cb10-14"><a href="#cb10-14" aria-hidden="true" tabindex="-1"></a>transitivity_df <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">Proteins =</span> <span class="fu">as.factor</span>(non_zero_trans_nodes),</span>
<span id="cb10-15"><a href="#cb10-15" aria-hidden="true" tabindex="-1"></a>                             <span class="at">Transitivity =</span> <span class="fu">as.numeric</span>(non_zero_trans))</span>
<span id="cb10-16"><a href="#cb10-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-17"><a href="#cb10-17" aria-hidden="true" tabindex="-1"></a><span class="co"># Barplot of non-zero transitivity values</span></span>
<span id="cb10-18"><a href="#cb10-18" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(transitivity_df, <span class="fu">aes</span>(<span class="at">x =</span>Proteins, <span class="at">y =</span> Transitivity))<span class="sc">+</span></span>
<span id="cb10-19"><a href="#cb10-19" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_bar</span>(<span class="at">stat =</span> <span class="st">"identity"</span>, <span class="at">fill =</span> <span class="st">"#495c51"</span>, <span class="at">width =</span> <span class="fl">0.8</span>)<span class="sc">+</span></span>
<span id="cb10-20"><a href="#cb10-20" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_minimal</span>()<span class="sc">+</span></span>
<span id="cb10-21"><a href="#cb10-21" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"Transitivity Values of Proteins"</span>,</span>
<span id="cb10-22"><a href="#cb10-22" aria-hidden="true" tabindex="-1"></a>       <span class="at">x =</span> <span class="st">"Nodes"</span>,</span>
<span id="cb10-23"><a href="#cb10-23" aria-hidden="true" tabindex="-1"></a>       <span class="at">y =</span> <span class="st">"Transitivity"</span>) <span class="sc">+</span></span>
<span id="cb10-24"><a href="#cb10-24" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">plot.title =</span> <span class="fu">element_text</span>(<span class="at">hjust=</span><span class="fl">0.5</span>),</span>
<span id="cb10-25"><a href="#cb10-25" aria-hidden="true" tabindex="-1"></a>        <span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">angle =</span> <span class="dv">90</span>, <span class="at">vjust =</span> <span class="fl">0.5</span>, <span class="at">hjust =</span> <span class="dv">1</span>, <span class="at">size =</span> <span class="dv">7</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div id="fig-transitivity" class="quarto-figure quarto-figure-center quarto-float anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-transitivity-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="BioNet_files/figure-html/fig-transitivity-1.png" class="img-fluid figure-img" width="672">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-transitivity-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;7: Barplot of transitivity values of nodes in PPI network.
</figcaption>
</figure>
</div>
</div>
</div>
<p>Transitivity values range from 0 to 1, with <span class="math inline">{\displaystyle C_T}</span> values close to 0 implying no closed triad for a given node, whereas <span class="math inline">{\displaystyle C_T}</span> values close or equal to 1 denote vertices that are part of completely closed triads. <a href="#fig-transitivity" class="quarto-xref">Figure&nbsp;7</a> shows that a small group of proteins in the network have low transitivity values, with most of the nodes having <span class="math inline">{\displaystyle C_T}</span> values above 0.5.</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Create table of APC transitivity value</span></span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a>APC_trans <span class="ot">&lt;-</span> <span class="fu">subset</span>(transitivity_df, Proteins <span class="sc">==</span> <span class="st">"APC"</span>)</span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a>APC_trans <span class="sc">|&gt;</span></span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(Proteins, Transitivity) <span class="sc">|&gt;</span></span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">gt</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div id="tbl-APC-transitivity" class="cell quarto-float anchored">
<figure class="quarto-float quarto-float-tbl figure">
<figcaption class="quarto-float-caption-top quarto-float-caption quarto-float-tbl" id="tbl-APC-transitivity-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Table&nbsp;1: Transitivity value of APC.
</figcaption>
<div aria-describedby="tbl-APC-transitivity-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<div class="cell-output-display">
<div id="ndfahktjog" style="padding-left:0px;padding-right:0px;padding-top:10px;padding-bottom:10px;overflow-x:auto;overflow-y:auto;width:auto;height:auto;">
<style>#ndfahktjog table {
  font-family: system-ui, 'Segoe UI', Roboto, Helvetica, Arial, sans-serif, 'Apple Color Emoji', 'Segoe UI Emoji', 'Segoe UI Symbol', 'Noto Color Emoji';
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
}

#ndfahktjog thead, #ndfahktjog tbody, #ndfahktjog tfoot, #ndfahktjog tr, #ndfahktjog td, #ndfahktjog th {
  border-style: none;
}

#ndfahktjog p {
  margin: 0;
  padding: 0;
}

#ndfahktjog .gt_table {
  display: table;
  border-collapse: collapse;
  line-height: normal;
  margin-left: auto;
  margin-right: auto;
  color: #333333;
  font-size: 16px;
  font-weight: normal;
  font-style: normal;
  background-color: #FFFFFF;
  width: auto;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #A8A8A8;
  border-right-style: none;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #A8A8A8;
  border-left-style: none;
  border-left-width: 2px;
  border-left-color: #D3D3D3;
}

#ndfahktjog .gt_caption {
  padding-top: 4px;
  padding-bottom: 4px;
}

#ndfahktjog .gt_title {
  color: #333333;
  font-size: 125%;
  font-weight: initial;
  padding-top: 4px;
  padding-bottom: 4px;
  padding-left: 5px;
  padding-right: 5px;
  border-bottom-color: #FFFFFF;
  border-bottom-width: 0;
}

#ndfahktjog .gt_subtitle {
  color: #333333;
  font-size: 85%;
  font-weight: initial;
  padding-top: 3px;
  padding-bottom: 5px;
  padding-left: 5px;
  padding-right: 5px;
  border-top-color: #FFFFFF;
  border-top-width: 0;
}

#ndfahktjog .gt_heading {
  background-color: #FFFFFF;
  text-align: center;
  border-bottom-color: #FFFFFF;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
}

#ndfahktjog .gt_bottom_border {
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
}

#ndfahktjog .gt_col_headings {
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
}

#ndfahktjog .gt_col_heading {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: normal;
  text-transform: inherit;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
  vertical-align: bottom;
  padding-top: 5px;
  padding-bottom: 6px;
  padding-left: 5px;
  padding-right: 5px;
  overflow-x: hidden;
}

#ndfahktjog .gt_column_spanner_outer {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: normal;
  text-transform: inherit;
  padding-top: 0;
  padding-bottom: 0;
  padding-left: 4px;
  padding-right: 4px;
}

#ndfahktjog .gt_column_spanner_outer:first-child {
  padding-left: 0;
}

#ndfahktjog .gt_column_spanner_outer:last-child {
  padding-right: 0;
}

#ndfahktjog .gt_column_spanner {
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  vertical-align: bottom;
  padding-top: 5px;
  padding-bottom: 5px;
  overflow-x: hidden;
  display: inline-block;
  width: 100%;
}

#ndfahktjog .gt_spanner_row {
  border-bottom-style: hidden;
}

#ndfahktjog .gt_group_heading {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  text-transform: inherit;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
  vertical-align: middle;
  text-align: left;
}

#ndfahktjog .gt_empty_group_heading {
  padding: 0.5px;
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  vertical-align: middle;
}

#ndfahktjog .gt_from_md > :first-child {
  margin-top: 0;
}

#ndfahktjog .gt_from_md > :last-child {
  margin-bottom: 0;
}

#ndfahktjog .gt_row {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  margin: 10px;
  border-top-style: solid;
  border-top-width: 1px;
  border-top-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 1px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 1px;
  border-right-color: #D3D3D3;
  vertical-align: middle;
  overflow-x: hidden;
}

#ndfahktjog .gt_stub {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  text-transform: inherit;
  border-right-style: solid;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
  padding-left: 5px;
  padding-right: 5px;
}

#ndfahktjog .gt_stub_row_group {
  color: #333333;
  background-color: #FFFFFF;
  font-size: 100%;
  font-weight: initial;
  text-transform: inherit;
  border-right-style: solid;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
  padding-left: 5px;
  padding-right: 5px;
  vertical-align: top;
}

#ndfahktjog .gt_row_group_first td {
  border-top-width: 2px;
}

#ndfahktjog .gt_row_group_first th {
  border-top-width: 2px;
}

#ndfahktjog .gt_summary_row {
  color: #333333;
  background-color: #FFFFFF;
  text-transform: inherit;
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
}

#ndfahktjog .gt_first_summary_row {
  border-top-style: solid;
  border-top-color: #D3D3D3;
}

#ndfahktjog .gt_first_summary_row.thick {
  border-top-width: 2px;
}

#ndfahktjog .gt_last_summary_row {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
}

#ndfahktjog .gt_grand_summary_row {
  color: #333333;
  background-color: #FFFFFF;
  text-transform: inherit;
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
}

#ndfahktjog .gt_first_grand_summary_row {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  border-top-style: double;
  border-top-width: 6px;
  border-top-color: #D3D3D3;
}

#ndfahktjog .gt_last_grand_summary_row_top {
  padding-top: 8px;
  padding-bottom: 8px;
  padding-left: 5px;
  padding-right: 5px;
  border-bottom-style: double;
  border-bottom-width: 6px;
  border-bottom-color: #D3D3D3;
}

#ndfahktjog .gt_striped {
  background-color: rgba(128, 128, 128, 0.05);
}

#ndfahktjog .gt_table_body {
  border-top-style: solid;
  border-top-width: 2px;
  border-top-color: #D3D3D3;
  border-bottom-style: solid;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
}

#ndfahktjog .gt_footnotes {
  color: #333333;
  background-color: #FFFFFF;
  border-bottom-style: none;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 2px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
}

#ndfahktjog .gt_footnote {
  margin: 0px;
  font-size: 90%;
  padding-top: 4px;
  padding-bottom: 4px;
  padding-left: 5px;
  padding-right: 5px;
}

#ndfahktjog .gt_sourcenotes {
  color: #333333;
  background-color: #FFFFFF;
  border-bottom-style: none;
  border-bottom-width: 2px;
  border-bottom-color: #D3D3D3;
  border-left-style: none;
  border-left-width: 2px;
  border-left-color: #D3D3D3;
  border-right-style: none;
  border-right-width: 2px;
  border-right-color: #D3D3D3;
}

#ndfahktjog .gt_sourcenote {
  font-size: 90%;
  padding-top: 4px;
  padding-bottom: 4px;
  padding-left: 5px;
  padding-right: 5px;
}

#ndfahktjog .gt_left {
  text-align: left;
}

#ndfahktjog .gt_center {
  text-align: center;
}

#ndfahktjog .gt_right {
  text-align: right;
  font-variant-numeric: tabular-nums;
}

#ndfahktjog .gt_font_normal {
  font-weight: normal;
}

#ndfahktjog .gt_font_bold {
  font-weight: bold;
}

#ndfahktjog .gt_font_italic {
  font-style: italic;
}

#ndfahktjog .gt_super {
  font-size: 65%;
}

#ndfahktjog .gt_footnote_marks {
  font-size: 75%;
  vertical-align: 0.4em;
  position: initial;
}

#ndfahktjog .gt_asterisk {
  font-size: 100%;
  vertical-align: 0;
}

#ndfahktjog .gt_indent_1 {
  text-indent: 5px;
}

#ndfahktjog .gt_indent_2 {
  text-indent: 10px;
}

#ndfahktjog .gt_indent_3 {
  text-indent: 15px;
}

#ndfahktjog .gt_indent_4 {
  text-indent: 20px;
}

#ndfahktjog .gt_indent_5 {
  text-indent: 25px;
}

#ndfahktjog .katex-display {
  display: inline-flex !important;
  margin-bottom: 0.75em !important;
}

#ndfahktjog div.Reactable > div.rt-table > div.rt-thead > div.rt-tr.rt-tr-group-header > div.rt-th-group:after {
  height: 0px !important;
}
</style>

<table class="gt_table do-not-create-environment cell caption-top table table-sm table-striped small" data-quarto-postprocess="true" data-quarto-disable-processing="false" data-quarto-bootstrap="false">
<thead>
<tr class="header gt_col_headings">
<th id="Proteins" class="gt_col_heading gt_columns_bottom_border gt_center" data-quarto-table-cell-role="th" scope="col">Proteins</th>
<th id="Transitivity" class="gt_col_heading gt_columns_bottom_border gt_right" data-quarto-table-cell-role="th" scope="col">Transitivity</th>
</tr>
</thead>
<tbody class="gt_table_body">
<tr class="odd">
<td class="gt_row gt_center" headers="Proteins">APC</td>
<td class="gt_row gt_right" headers="Transitivity">0.0009493671</td>
</tr>
</tbody>
</table>

</div>
</div>
</div>
</figure>
</div>
</div>
<p><a href="#fig-transitivity" class="quarto-xref">Figure&nbsp;7</a> shows that APC, the protein that has the highest betweenness centrality value of all other vertices in the network, exhibits the lowest possible value of transitivity. More specifically, APC has a transitivity value of about 0.00095, as shown in <a href="#tbl-APC-transitivity" class="quarto-xref">Table&nbsp;1</a>. This possibly constitutes a counter-intuitive combination for a node; having both high betweenness centrality and substantially low transitivity values. However, a closer look into the community’s structure and how the nodes are organized around APC will elucidate the two statistical measures and their meanings.</p>
</section>
</section>
<section id="the-importance-of-hubs" class="level3" data-number="2.4">
<h3 data-number="2.4" class="anchored" data-anchor-id="the-importance-of-hubs"><span class="header-section-number">2.4</span> The Importance Of Hubs</h3>
<p>Focusing on a community can be helpful when it comes to looking into possible important constituents of a network. The close examination of the interactions of proteins in a particular community can lead to the identification of protein hubs, proteins that interact with many others within the cell and often play a crucial role in the organization and regulation of PPI networks, being essential for the stability and functionality of cellular processes. Many hub proteins are evolutionarily conserved, indicating their essential role across different species <span class="citation" data-cites="vallabhajosyula2009a">(<a href="#ref-vallabhajosyula2009a" role="doc-biblioref">Vallabhajosyula et al. 2009</a>)</span>.</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Assign the community APC is a part of to a variable </span></span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a>target_community <span class="ot">&lt;-</span> <span class="fu">V</span>(tbl_graph)<span class="sc">$</span>community[<span class="fu">V</span>(tbl_graph)<span class="sc">$</span>name <span class="sc">==</span> <span class="st">"APC"</span>]</span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Create vector with nodes that belong to the community APC belongs to</span></span>
<span id="cb12-5"><a href="#cb12-5" aria-hidden="true" tabindex="-1"></a>nodes_in_community <span class="ot">&lt;-</span> <span class="fu">which</span>(membership <span class="sc">==</span> target_community)</span>
<span id="cb12-6"><a href="#cb12-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-7"><a href="#cb12-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Create subgraph object of APC community nodes and turn it into tbl_graph object</span></span>
<span id="cb12-8"><a href="#cb12-8" aria-hidden="true" tabindex="-1"></a>subgraph1 <span class="ot">&lt;-</span> <span class="fu">subgraph</span>(graph, nodes_in_community)</span>
<span id="cb12-9"><a href="#cb12-9" aria-hidden="true" tabindex="-1"></a>tbl_subgraph1 <span class="ot">&lt;-</span> <span class="fu">as_tbl_graph</span>(subgraph1)</span>
<span id="cb12-10"><a href="#cb12-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-11"><a href="#cb12-11" aria-hidden="true" tabindex="-1"></a><span class="co"># Add degree attribute for vertices in tbl_subgraph1</span></span>
<span id="cb12-12"><a href="#cb12-12" aria-hidden="true" tabindex="-1"></a><span class="fu">V</span>(tbl_subgraph1)<span class="sc">$</span>Degree <span class="ot">&lt;-</span> <span class="fu">degree</span>(subgraph1)</span>
<span id="cb12-13"><a href="#cb12-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-14"><a href="#cb12-14" aria-hidden="true" tabindex="-1"></a><span class="co"># APC Community plot with node size according to node degree</span></span>
<span id="cb12-15"><a href="#cb12-15" aria-hidden="true" tabindex="-1"></a><span class="fu">ggraph</span>(tbl_subgraph1, <span class="at">layout =</span> <span class="st">"fr"</span>) <span class="sc">+</span> </span>
<span id="cb12-16"><a href="#cb12-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_edge_link</span>(<span class="at">alpha =</span> <span class="fl">0.8</span>, <span class="at">color =</span> <span class="st">"darkgrey"</span>) <span class="sc">+</span> </span>
<span id="cb12-17"><a href="#cb12-17" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_node_point</span>(<span class="fu">aes</span>(<span class="at">size =</span> <span class="fu">V</span>(tbl_subgraph1)<span class="sc">$</span>Degree), <span class="at">shape =</span> <span class="dv">21</span>, <span class="at">border=</span><span class="st">"black"</span>,</span>
<span id="cb12-18"><a href="#cb12-18" aria-hidden="true" tabindex="-1"></a>                  <span class="at">fill=</span><span class="st">"#174e50"</span>, <span class="at">stroke =</span> <span class="fl">0.1</span>) <span class="sc">+</span></span>
<span id="cb12-19"><a href="#cb12-19" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_node_text</span>(<span class="fu">aes</span>(<span class="at">label=</span>name), <span class="at">vjust =</span> <span class="fl">1.5</span>, <span class="at">hjust =</span> <span class="dv">1</span>, <span class="at">size =</span> <span class="fl">2.3</span>) <span class="sc">+</span></span>
<span id="cb12-20"><a href="#cb12-20" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_void</span>() <span class="sc">+</span></span>
<span id="cb12-21"><a href="#cb12-21" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"APC Community"</span>) <span class="sc">+</span></span>
<span id="cb12-22"><a href="#cb12-22" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">plot.title =</span> <span class="fu">element_text</span>(<span class="at">hjust =</span> <span class="fl">0.5</span>),</span>
<span id="cb12-23"><a href="#cb12-23" aria-hidden="true" tabindex="-1"></a>        <span class="at">plot.margin =</span> <span class="fu">margin</span>(<span class="at">t =</span> <span class="dv">40</span>, <span class="at">r =</span> <span class="dv">40</span>, <span class="at">b =</span> <span class="dv">40</span>, <span class="at">l =</span> <span class="dv">40</span>),</span>
<span id="cb12-24"><a href="#cb12-24" aria-hidden="true" tabindex="-1"></a>        <span class="at">legend.position =</span> <span class="st">"none"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div id="fig-APC-community" class="quarto-figure quarto-figure-center quarto-float anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-APC-community-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="BioNet_files/figure-html/fig-APC-community-1.png" class="img-fluid figure-img" width="672">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-APC-community-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;8: Graph of nodes belonging to the same community as APC. Node sizes are proportionate to node degrees.
</figcaption>
</figure>
</div>
</div>
</div>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Find neighbors of APC and assign them to vector</span></span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a>APC_neighbors <span class="ot">&lt;-</span> <span class="fu">neighbors</span>(subgraph1, <span class="fu">V</span>(subgraph1)<span class="sc">$</span>name <span class="sc">==</span> <span class="st">"APC"</span>)</span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Assign APC neighbors' names to vector</span></span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true" tabindex="-1"></a>APC_net <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"APC"</span>, <span class="fu">V</span>(subgraph1)<span class="sc">$</span>name[APC_neighbors])</span>
<span id="cb13-6"><a href="#cb13-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-7"><a href="#cb13-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Create subgraph of APC neighbors and turn into tbl_graph object</span></span>
<span id="cb13-8"><a href="#cb13-8" aria-hidden="true" tabindex="-1"></a>APC_net_graph <span class="ot">&lt;-</span> <span class="fu">subgraph</span>(subgraph1, APC_net)</span>
<span id="cb13-9"><a href="#cb13-9" aria-hidden="true" tabindex="-1"></a>tbl_APC <span class="ot">&lt;-</span> <span class="fu">as_tbl_graph</span>(APC_net_graph)</span>
<span id="cb13-10"><a href="#cb13-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-11"><a href="#cb13-11" aria-hidden="true" tabindex="-1"></a><span class="co"># APC neighbors plot</span></span>
<span id="cb13-12"><a href="#cb13-12" aria-hidden="true" tabindex="-1"></a><span class="fu">ggraph</span>(tbl_APC, <span class="at">layout =</span> <span class="st">"fr"</span>) <span class="sc">+</span> </span>
<span id="cb13-13"><a href="#cb13-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_edge_link</span>(<span class="at">alpha =</span> <span class="fl">0.8</span>, <span class="at">color =</span> <span class="st">"darkgrey"</span>) <span class="sc">+</span> </span>
<span id="cb13-14"><a href="#cb13-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_node_point</span>(<span class="fu">aes</span>(<span class="at">size =</span> <span class="dv">4</span>), <span class="at">shape =</span> <span class="dv">21</span>, <span class="at">border=</span><span class="st">"black"</span>,</span>
<span id="cb13-15"><a href="#cb13-15" aria-hidden="true" tabindex="-1"></a>                  <span class="at">fill=</span><span class="st">"#174e50"</span>, <span class="at">stroke =</span> <span class="fl">0.1</span>) <span class="sc">+</span></span>
<span id="cb13-16"><a href="#cb13-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_node_text</span>(<span class="fu">aes</span>(<span class="at">label=</span>name), <span class="at">vjust =</span> <span class="dv">2</span>, <span class="at">hjust =</span> <span class="fl">0.5</span>, <span class="at">size =</span> <span class="fl">2.3</span>) <span class="sc">+</span></span>
<span id="cb13-17"><a href="#cb13-17" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_void</span>() <span class="sc">+</span></span>
<span id="cb13-18"><a href="#cb13-18" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"APC Neighbors"</span>) <span class="sc">+</span></span>
<span id="cb13-19"><a href="#cb13-19" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">plot.title =</span> <span class="fu">element_text</span>(<span class="at">hjust =</span> <span class="fl">0.5</span>),</span>
<span id="cb13-20"><a href="#cb13-20" aria-hidden="true" tabindex="-1"></a>        <span class="at">plot.margin =</span> <span class="fu">margin</span>(<span class="at">t =</span> <span class="dv">40</span>, <span class="at">r =</span> <span class="dv">40</span>, <span class="at">b =</span> <span class="dv">40</span>, <span class="at">l =</span> <span class="dv">40</span>),</span>
<span id="cb13-21"><a href="#cb13-21" aria-hidden="true" tabindex="-1"></a>        <span class="at">legend.position =</span> <span class="st">"none"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div id="fig-APC-neighbors" class="quarto-figure quarto-figure-center quarto-float anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-APC-neighbors-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="BioNet_files/figure-html/fig-APC-neighbors-1.png" class="img-fluid figure-img" width="672">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-APC-neighbors-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;9: Graph of APC neighbors.
</figcaption>
</figure>
</div>
</div>
</div>
<p><a href="#fig-APC-community" class="quarto-xref">Figure&nbsp;8</a> depicts the community of APC and its neighboring nodes. It is clear that APC is a central node and has numerous connections to the rest of the vertices in the community, which hints to the high degree and betweenness centrality of the protein in question. In order to clearly display what a low transitivity value can be translated to one needs to look into the neighboring nodes of APC, that is, nodes that are connected to the protein with a direct link. As shown in <a href="#fig-APC-neighbors" class="quarto-xref">Figure&nbsp;9</a>, the neighbors of APC are hardly connected to each other, which indicates low degree of triad formation or local clustering in the community.</p>
<p>It is important to highlight the fact that the central position, high connectivity, node degree and betweenness centrality of APC constitute characteristic properties of hub nodes. In other words, APC, being one of the few central vertices of the system that has lots of connections, contributes to the increase of the network’s topological robustness and flow of information. This example demonstrates the fact that some highly connected hubs can have low transitivity values, linking otherwise not communicating nodes.</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Assign the community ATF7IP is a part of to a variable </span></span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a>target_com <span class="ot">&lt;-</span> <span class="fu">V</span>(tbl_graph)<span class="sc">$</span>community[<span class="fu">V</span>(tbl_graph)<span class="sc">$</span>name <span class="sc">==</span> <span class="st">"ATF7IP"</span>]</span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Create vector with nodes that belong to the community ATF7IP belongs to</span></span>
<span id="cb14-5"><a href="#cb14-5" aria-hidden="true" tabindex="-1"></a>nodes_in_com <span class="ot">&lt;-</span> <span class="fu">which</span>(membership <span class="sc">==</span> target_com)</span>
<span id="cb14-6"><a href="#cb14-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-7"><a href="#cb14-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Create subgraph object of ATF7IP community nodes and turn it into tbl_graph object</span></span>
<span id="cb14-8"><a href="#cb14-8" aria-hidden="true" tabindex="-1"></a>subgraph2 <span class="ot">&lt;-</span> <span class="fu">subgraph</span>(graph, nodes_in_com)</span>
<span id="cb14-9"><a href="#cb14-9" aria-hidden="true" tabindex="-1"></a>tbl_subgraph2 <span class="ot">&lt;-</span> <span class="fu">as_tbl_graph</span>(subgraph2)</span>
<span id="cb14-10"><a href="#cb14-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-11"><a href="#cb14-11" aria-hidden="true" tabindex="-1"></a><span class="co"># Add degree attribute for vertices in tbl_subgraph2</span></span>
<span id="cb14-12"><a href="#cb14-12" aria-hidden="true" tabindex="-1"></a><span class="fu">V</span>(tbl_subgraph2)<span class="sc">$</span>Degree <span class="ot">&lt;-</span> <span class="fu">degree</span>(subgraph2)</span>
<span id="cb14-13"><a href="#cb14-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-14"><a href="#cb14-14" aria-hidden="true" tabindex="-1"></a><span class="co"># ATF7IP Community plot with node size according to node degree</span></span>
<span id="cb14-15"><a href="#cb14-15" aria-hidden="true" tabindex="-1"></a><span class="fu">ggraph</span>(tbl_subgraph2, <span class="at">layout =</span> <span class="st">"fr"</span>) <span class="sc">+</span> </span>
<span id="cb14-16"><a href="#cb14-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_edge_link</span>(<span class="at">alpha =</span> <span class="fl">0.8</span>, <span class="at">color =</span> <span class="st">"darkgrey"</span>) <span class="sc">+</span> </span>
<span id="cb14-17"><a href="#cb14-17" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_node_point</span>(<span class="fu">aes</span>(<span class="at">size =</span> <span class="fu">V</span>(tbl_subgraph2)<span class="sc">$</span>Degree), <span class="at">shape =</span> <span class="dv">21</span>, <span class="at">border=</span><span class="st">"black"</span>,</span>
<span id="cb14-18"><a href="#cb14-18" aria-hidden="true" tabindex="-1"></a>                  <span class="at">fill=</span><span class="st">"#174e50"</span>, <span class="at">stroke =</span> <span class="fl">0.1</span>) <span class="sc">+</span></span>
<span id="cb14-19"><a href="#cb14-19" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_node_text</span>(<span class="fu">aes</span>(<span class="at">label=</span>name), <span class="at">vjust =</span> <span class="fl">1.8</span>, <span class="at">hjust =</span> <span class="fl">0.5</span>, <span class="at">size =</span> <span class="fl">2.3</span>) <span class="sc">+</span></span>
<span id="cb14-20"><a href="#cb14-20" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_void</span>() <span class="sc">+</span></span>
<span id="cb14-21"><a href="#cb14-21" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"ATF7IP Community"</span>) <span class="sc">+</span></span>
<span id="cb14-22"><a href="#cb14-22" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">plot.title =</span> <span class="fu">element_text</span>(<span class="at">hjust =</span> <span class="fl">0.5</span>),</span>
<span id="cb14-23"><a href="#cb14-23" aria-hidden="true" tabindex="-1"></a>        <span class="at">plot.margin =</span> <span class="fu">margin</span>(<span class="at">t =</span> <span class="dv">40</span>, <span class="at">r =</span> <span class="dv">40</span>, <span class="at">b =</span> <span class="dv">40</span>, <span class="at">l =</span> <span class="dv">40</span>),</span>
<span id="cb14-24"><a href="#cb14-24" aria-hidden="true" tabindex="-1"></a>        <span class="at">legend.position =</span> <span class="st">"none"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div id="fig-ATF7IP-community" class="quarto-figure quarto-figure-center quarto-float anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-ATF7IP-community-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="BioNet_files/figure-html/fig-ATF7IP-community-1.png" class="img-fluid figure-img" width="672">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-ATF7IP-community-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;10: Graph of nodes belonging to the same community as ATF7IP. Node sizes are proportionate to node degrees.
</figcaption>
</figure>
</div>
</div>
</div>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Find neighbors of ATF7IP and assign them to vector</span></span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a>ATF7IP_neighbors <span class="ot">&lt;-</span> <span class="fu">neighbors</span>(subgraph2, <span class="fu">V</span>(subgraph2)<span class="sc">$</span>name <span class="sc">==</span> <span class="st">"ATF7IP"</span>)</span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Assign ATF7IP neighbors' names to vector</span></span>
<span id="cb15-5"><a href="#cb15-5" aria-hidden="true" tabindex="-1"></a>ATF7IP_net <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"ATF7IP"</span>, <span class="fu">V</span>(subgraph2)<span class="sc">$</span>name[ATF7IP_neighbors])</span>
<span id="cb15-6"><a href="#cb15-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-7"><a href="#cb15-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Create subgraph of ATF7IP neighbors and turn into tbl_graph object</span></span>
<span id="cb15-8"><a href="#cb15-8" aria-hidden="true" tabindex="-1"></a>ATF7IP_net_graph <span class="ot">&lt;-</span> <span class="fu">subgraph</span>(subgraph2, ATF7IP_net)</span>
<span id="cb15-9"><a href="#cb15-9" aria-hidden="true" tabindex="-1"></a>tbl_ATF <span class="ot">&lt;-</span> <span class="fu">as_tbl_graph</span>(ATF7IP_net_graph)</span>
<span id="cb15-10"><a href="#cb15-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-11"><a href="#cb15-11" aria-hidden="true" tabindex="-1"></a><span class="co"># ATF7IP neighbors plot</span></span>
<span id="cb15-12"><a href="#cb15-12" aria-hidden="true" tabindex="-1"></a><span class="fu">ggraph</span>(tbl_ATF, <span class="at">layout =</span> <span class="st">"fr"</span>) <span class="sc">+</span> </span>
<span id="cb15-13"><a href="#cb15-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_edge_link</span>(<span class="at">alpha =</span> <span class="fl">0.8</span>, <span class="at">color =</span> <span class="st">"darkgrey"</span>) <span class="sc">+</span> </span>
<span id="cb15-14"><a href="#cb15-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_node_point</span>(<span class="fu">aes</span>(<span class="at">size =</span> <span class="dv">4</span>), <span class="at">shape =</span> <span class="dv">21</span>, <span class="at">border=</span><span class="st">"black"</span>,</span>
<span id="cb15-15"><a href="#cb15-15" aria-hidden="true" tabindex="-1"></a>                  <span class="at">fill=</span><span class="st">"#174e50"</span>, <span class="at">stroke =</span> <span class="fl">0.1</span>) <span class="sc">+</span></span>
<span id="cb15-16"><a href="#cb15-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_node_text</span>(<span class="fu">aes</span>(<span class="at">label=</span>name), <span class="at">vjust =</span> <span class="fl">1.8</span>, <span class="at">hjust =</span> <span class="fl">0.5</span>, <span class="at">size =</span> <span class="fl">2.3</span>) <span class="sc">+</span></span>
<span id="cb15-17"><a href="#cb15-17" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_void</span>() <span class="sc">+</span></span>
<span id="cb15-18"><a href="#cb15-18" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"ATF7IP Neighbors"</span>) <span class="sc">+</span></span>
<span id="cb15-19"><a href="#cb15-19" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">plot.title =</span> <span class="fu">element_text</span>(<span class="at">hjust =</span> <span class="fl">0.5</span>),</span>
<span id="cb15-20"><a href="#cb15-20" aria-hidden="true" tabindex="-1"></a>        <span class="at">plot.margin =</span> <span class="fu">margin</span>(<span class="at">t =</span> <span class="dv">40</span>, <span class="at">r =</span> <span class="dv">40</span>, <span class="at">b =</span> <span class="dv">40</span>, <span class="at">l =</span> <span class="dv">40</span>),</span>
<span id="cb15-21"><a href="#cb15-21" aria-hidden="true" tabindex="-1"></a>        <span class="at">legend.position =</span> <span class="st">"none"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div id="fig-ATF7IP-neighbors" class="quarto-figure quarto-figure-center quarto-float anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-ATF7IP-neighbors-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="BioNet_files/figure-html/fig-ATF7IP-neighbors-1.png" class="img-fluid figure-img" width="672">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-ATF7IP-neighbors-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;11: Graph of ATF7IP neighbors.
</figcaption>
</figure>
</div>
</div>
</div>
<p>For comparison purposes, the same analysis was done for one of the proteins with a transitivity value equal to 1, ATF7IP. <a href="#fig-ATF7IP-community" class="quarto-xref">Figure&nbsp;10</a> displays the community that ATF7IP is a member of, showcasing the relatively low node degree of the protein, as well as the vastly different structure of the community, conveying more complex relationships between nodes, when compared to <a href="#fig-APC-community" class="quarto-xref">Figure&nbsp;8</a>. The high transitivity value of ATF7IP is shown in <a href="#fig-ATF7IP-neighbors" class="quarto-xref">Figure&nbsp;11</a>, where it is clear that all the possible triads that could be formed between ATF7IP’s neighboring vertices already exist in the small group.</p>
</section>
</section>
<section id="conclusion" class="level2" data-number="3">
<h2 data-number="3" class="anchored" data-anchor-id="conclusion"><span class="header-section-number">3</span> Conclusion</h2>
<p>Network Biology is an interdisciplinary field of research that can be of immense importance for the deeper understanding of biological systems, including protein-protein interaction networks. Network theory provides the tools for the study of such complex systems, using a variety of statistical measures, including node degree, betweenness centrality, transitivity and modularity to distinguish between different types of networks and their distinct behavior and evolution. These types of analyses can be a starting point for more detailed experimental studies to explore these interactions and their functional implications.</p>

</section>

<div id="quarto-appendix" class="default"><section class="quarto-appendix-contents" role="doc-bibliography" id="quarto-bibliography"><h2 class="anchored quarto-appendix-heading">References</h2><div id="refs" class="references csl-bib-body hanging-indent" data-entry-spacing="0" role="list">
<div id="ref-abbott2023" class="csl-entry" role="listitem">
Abbott, James, and Inke S. Näthke. 2023. <span>“The Adenomatous Polyposis Coli Protein 30 Years On.”</span> <em>Seminars in Cell &amp; Developmental Biology</em> 150-151 (December): 28–34. <a href="https://doi.org/10.1016/j.semcdb.2023.04.004">https://doi.org/10.1016/j.semcdb.2023.04.004</a>.
</div>
<div id="ref-albert2000" class="csl-entry" role="listitem">
Albert, Réka, Hawoong Jeong, and Albert-László Barabási. 2000. <span>“Error and Attack Tolerance of Complex Networks.”</span> <em>Nature</em> 406 (6794): 378–82. <a href="https://doi.org/10.1038/35019019">https://doi.org/10.1038/35019019</a>.
</div>
<div id="ref-bandyopadhyay2010" class="csl-entry" role="listitem">
Bandyopadhyay, Sourav, Monika Mehta, Dwight Kuo, Min-Kyung Sung, Ryan Chuang, Eric J. Jaehnig, Bernd Bodenmiller, et al. 2010. <span>“Rewiring of Genetic Networks in Response to DNA Damage.”</span> <em>Science</em> 330 (6009): 1385–89. <a href="https://doi.org/10.1126/science.1195618">https://doi.org/10.1126/science.1195618</a>.
</div>
<div id="ref-barabási2004" class="csl-entry" role="listitem">
Barabási, Albert-László, and Zoltán N. Oltvai. 2004. <span>“Network Biology: Understanding the Cell’s Functional Organization.”</span> <em>Nature Reviews Genetics</em> 5 (2): 101–13. <a href="https://doi.org/10.1038/nrg1272">https://doi.org/10.1038/nrg1272</a>.
</div>
<div id="ref-blondel2008" class="csl-entry" role="listitem">
Blondel, Vincent D, Jean-Loup Guillaume, Renaud Lambiotte, and Etienne Lefebvre. 2008. <span>“Fast Unfolding of Communities in Large Networks.”</span> <em>Journal of Statistical Mechanics: Theory and Experiment</em> 2008 (10): P10008. <a href="https://doi.org/10.1088/1742-5468/2008/10/p10008">https://doi.org/10.1088/1742-5468/2008/10/p10008</a>.
</div>
<div id="ref-liu2017" class="csl-entry" role="listitem">
Liu, Jing, Mingxing Zhou, Shuai Wang, and Penghui Liu. 2017. <span>“A Comparative Study of Network Robustness Measures.”</span> <em>Frontiers of Computer Science</em> 11 (4): 568–84. <a href="https://doi.org/10.1007/s11704-016-6108-z">https://doi.org/10.1007/s11704-016-6108-z</a>.
</div>
<div id="ref-newman2006" class="csl-entry" role="listitem">
Newman, M. E. J. 2006. <span>“Modularity and Community Structure in Networks.”</span> <em>Proceedings of the National Academy of Sciences</em> 103 (23): 8577–82. <a href="https://doi.org/10.1073/pnas.0601602103">https://doi.org/10.1073/pnas.0601602103</a>.
</div>
<div id="ref-opsahl2010" class="csl-entry" role="listitem">
Opsahl, Tore, Filip Agneessens, and John Skvoretz. 2010. <span>“Node Centrality in Weighted Networks: Generalizing Degree and Shortest Paths.”</span> <em>Social Networks</em> 32 (3): 245–51. <a href="https://doi.org/10.1016/j.socnet.2010.03.006">https://doi.org/10.1016/j.socnet.2010.03.006</a>.
</div>
<div id="ref-przulj2004" class="csl-entry" role="listitem">
Pržulj, N., D. G. Corneil, and I. Jurisica. 2004. <span>“Modeling Interactome: Scale-Free or Geometric?”</span> <em>Bioinformatics</em> 20 (18): 3508–15. <a href="https://doi.org/10.1093/bioinformatics/bth436">https://doi.org/10.1093/bioinformatics/bth436</a>.
</div>
<div id="ref-suter2008" class="csl-entry" role="listitem">
Suter, Bernhard, Saranya Kittanakom, and Igor Stagljar. 2008. <span>“Two-Hybrid Technologies in Proteomics Research.”</span> <em>Current Opinion in Biotechnology</em> 19 (4): 316–23. <a href="https://doi.org/10.1016/j.copbio.2008.06.005">https://doi.org/10.1016/j.copbio.2008.06.005</a>.
</div>
<div id="ref-vallabhajosyula2009a" class="csl-entry" role="listitem">
Vallabhajosyula, Ravishankar R., Deboki Chakravarti, Samina Lutfeali, Animesh Ray, and Alpan Raval. 2009. <span>“Identifying Hubs in Protein Interaction Networks.”</span> Edited by Diego Di Bernardo. <em>PLoS ONE</em> 4 (4): e5344. <a href="https://doi.org/10.1371/journal.pone.0005344">https://doi.org/10.1371/journal.pone.0005344</a>.
</div>
</div></section></div></main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  const viewSource = window.document.getElementById('quarto-view-source') ||
                     window.document.getElementById('quarto-code-tools-source');
  if (viewSource) {
    const sourceUrl = viewSource.getAttribute("data-quarto-source-url");
    viewSource.addEventListener("click", function(e) {
      if (sourceUrl) {
        // rstudio viewer pane
        if (/\bcapabilities=\b/.test(window.location)) {
          window.open(sourceUrl);
        } else {
          window.location.href = sourceUrl;
        }
      } else {
        const modal = new bootstrap.Modal(document.getElementById('quarto-embedded-source-code-modal'));
        modal.show();
      }
      return false;
    });
  }
  function toggleCodeHandler(show) {
    return function(e) {
      const detailsSrc = window.document.querySelectorAll(".cell > details > .sourceCode");
      for (let i=0; i<detailsSrc.length; i++) {
        const details = detailsSrc[i].parentElement;
        if (show) {
          details.open = true;
        } else {
          details.removeAttribute("open");
        }
      }
      const cellCodeDivs = window.document.querySelectorAll(".cell > .sourceCode");
      const fromCls = show ? "hidden" : "unhidden";
      const toCls = show ? "unhidden" : "hidden";
      for (let i=0; i<cellCodeDivs.length; i++) {
        const codeDiv = cellCodeDivs[i];
        if (codeDiv.classList.contains(fromCls)) {
          codeDiv.classList.remove(fromCls);
          codeDiv.classList.add(toCls);
        } 
      }
      return false;
    }
  }
  const hideAllCode = window.document.getElementById("quarto-hide-all-code");
  if (hideAllCode) {
    hideAllCode.addEventListener("click", toggleCodeHandler(false));
  }
  const showAllCode = window.document.getElementById("quarto-show-all-code");
  if (showAllCode) {
    showAllCode.addEventListener("click", toggleCodeHandler(true));
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->




</body></html>